core/resources.rs

```diff
@@ -27,6 +34,21 @@ pub trait Resource: Any + 'static {
     type_name::<Self>().into()
   }
 
+  /// Resources may implement `read()` to be a readable stream
+  fn read(self: Rc<Self>, _buf: ZeroCopyBuf) -> Option<AsyncResult<usize>> {
```

Comment:
It is possible to make this more ergonomic by using the [async_trait] proc macro (which btw also boxes the promise returned by an async function, hidden from view)

Reply:
I had considered it, but thought it was best to avoid the extra dep and "macro magic" since there's only a handful of call sites that shouldn't frequently change